### ISSUE TITLE
Proposal: Run Provider Verification as subtests

### ISSUE THREAD

**Issue by mefellows** [pact-foundation/pact-go#58]
This is a thought bubble for a potential feature.

Thanks to #57, we now have granular access to all of the failures in a particular Provider verification. 

Prior to this, all failures for a Provider were wrapped into a single test case, which is quite annoying if you have a large volume of tests and want to see in a report (often in a CI server) exactly where something failed.

Conversely, it is also nice to know how many tests were actually performed as a metric, something which is obscured in the current interface.

One suggestion could be to modify or overload the `VerifyProvider` to accept a parameter of type `testing.T` which records all interactions as [https://blog.golang.org/subtests](subtests). 

A typical Provider side test would like something like:

	func TestProvider_PactContract(t *testing.T) {
                // Create Pact, connecting to local Daemon
	        // Ensure the port matches the daemon port!
		pact := Pact{
			Port: 6666,
		}
		go startMyAPI("http://localhost:8000")

		pact.VerifyProvider(t, types.VerifyRequest{
			ProviderBaseURL:        "http://localhost:8000",
			PactURLs:               []string{"./pacts/my_consumer-my_provider.json"},
			ProviderStatesSetupURL: "http://localhost:8000/setup",
		})
	}

Behind the scenes, this will automatically fail the test if there is an error, and record all of the verifications as subtests, giving us something like the following:

```
--- FAIL: TestProvider_PactContract (0.00s)
    --- PASS:TestProvider_PactContract_GetHello (0.00s)
    --- FAIL: TestProvider_PactContract_GetHello (0.00s)
        provider_test.go:84: expected 200 but got 404
```

**Comment by mefellows** [comment]
cc: @bmarini keen to get your thoughts

**Comment by mefellows** [comment]
Just whipped together a quick failure to see how it comes out, here is a broken Gin example:


```
--- FAIL: TestPact_Provider (0.56s)
    --- FAIL: TestPact_Provider/has_status_code_404 (0.00s)
    	pact.go:269: Verifying a pact between billy and bobby Given User billy does not exist A request to login with user 'billy' with POST /users/login returns a response which has status code 404
    	pact.go:271:
    		expected: 404
    		     got: 200

    		(compared using eql?)

    --- FAIL: TestPact_Provider/has_status_code_401 (0.00s)
    	pact.go:269: Verifying a pact between billy and bobby Given User billy is unauthorized A request to login with user 'billy' with POST /users/login returns a response which has status code 401
    	pact.go:271:
    		expected: 401
    		     got: 200

    		(compared using eql?)

FAIL
FAIL	github.com/pact-foundation/pact-go/examples/gin/provider	0.577s
```

**Comment by mefellows** [comment]
Closed by #59
